### PR TITLE
Laboratory work #3, Andrey Gusev - 22FPL1

### PR DESCRIPTION
**Pull request by AndreyGusev1**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def __init__(self, end_of_word_token: str) -> None:
         Args:
             end_of_word_token (str): A token denoting word boundary
         """
+        self._end_of_word_token = '_'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:26]: strange constant

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,17 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str | None):
+            return None
+        tokenized_text = []
+        for word in text.lower().split():
+            for letter in word:
+                if letter.isalpha():
+                    tokenized_text.append(letter)
+            tokenized_text.append('_')
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:53]: again

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (AndreyGusev1) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (AndreyGusev1) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -159,6 +220,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        result = ''
+        for i, j in enumerate(decoded_corpus):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:226]: Better names, please

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (AndreyGusev1) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or len(text) == 0:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:46]: simplify this: `len(text)==0`

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+        tokenized_text = []
+        counter = 0
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:49]: Wow... check it simply

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -76,6 +97,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+        for keys, values in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:102]: filter

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (AndreyGusev1) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -41,6 +43,19 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or not text:
+            return None
+        tokenized_text = []
+        for word in text.lower():
+            if word.isalpha():
+                tokenized_text.append(word)
+            elif word is ' ' and tokenized_text[-1] != self._end_of_word_token:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:52]: just `not word`

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -159,6 +218,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        result = ''
+        for index, symbol in enumerate(decoded_corpus):
+            if symbol == self._end_of_word_token and index == len(decoded_corpus)-1:
+                result += '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:226]: fstring

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (AndreyGusev1) [reply lab_3_generate_by_ngrams/main.py]: fixed

**Pull request closed without merging**